### PR TITLE
chore(warehouse): wh-test for configuration testing for rudder-cli

### PR DESCRIPTION
**Pull request by achettyiitr**
# Description

> Provide a way to run warehouse configuration test through rudder-cli.

## Notion Ticket

> https://www.notion.so/rudderstacks/Warehouse-configuration-testing-through-rudder-cli-6c52fc89a6db473f87693fb315766f61

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: warehouse/admin.go [hunk 1]
```diff
@@ -14,6 +16,21 @@ import (
 
 type WarehouseAdmin struct{}
 
+type QueryInput struct {
+	DestID       string
+	SourceID     string
+	SQLStatement string
+}
+
+type TestInput struct {
```
Comment (abhimanyubabbar) [inline warehouse/admin.go:25]: Let's name it `DestinationValidationInput` ?

File: warehouse/admin.go [hunk 2]
```diff
@@ -14,6 +16,21 @@ import (
...
+	DestID string
+}
+
+type TestResponse struct {
```
Comment (abhimanyubabbar) [inline warehouse/admin.go:29]: `DestinationValidationOutput` ?

File: warehouse/admin.go [hunk 3]
```diff
@@ -77,30 +88,34 @@ func (wh *WarehouseAdmin) Query(s QueryInput, reply *warehouseutils.QueryResult)
...
 
-func (wh *WarehouseAdmin) QueryWhUploads(uploadsReq UploadsReqT, reply *[]byte) error {
-	uploadsReq.API = UploadAPI
-	res, err := uploadsReq.GetWhUploads()
-	if err != nil {
-		return err
+// Test the underlying warehouse
+func (wh *WarehouseAdmin) Test(s TestInput, reply *TestResponse) (err error) {
+	if strings.TrimSpace(s.DestID) == "" {
+		return errors.New("please specify the destination ID to query the warehouse")
```
Comment (abhimanyubabbar) [inline warehouse/admin.go:94]: I see you use named return arguments extensively in codebase. I was thinking about them a lot and it has good benefits but one thing it causes is mixed return statements and implicit shadowing of variables.

Let's discuss this 1-1 when you have time ?

File: warehouse/admin.go [hunk 4]
```diff
@@ -77,30 +88,34 @@ func (wh *WarehouseAdmin) Query(s QueryInput, reply *warehouseutils.QueryResult)
...
 	}
-	bytes, err := json.Marshal(res)
-	if err != nil {
-		return err
+
+	var warehouse warehouseutils.WarehouseT
+	srcMap, ok := connectionsMap[s.DestID]
+	if !ok {
+		return errors.New("please specify a valid and existing destination ID")
```
Comment (abhimanyubabbar) [inline warehouse/admin.go:100]: Add the destinationID in the error response, so we can log it appropriately. Wdyt ?

File: warehouse/admin.go [hunk 5]
```diff
@@ -77,30 +88,34 @@ func (wh *WarehouseAdmin) Query(s QueryInput, reply *warehouseutils.QueryResult)
...
 	if err != nil {
-		return err
+		err = fmt.Errorf("unable to successfully validate destination: %s credentials, err: %v", warehouse.Destination.ID, err)
+		pkgLogger.Error(err)
+		return
 	}
-	*reply = bytes
-	return nil
+
+	*reply = TestResponse{Status: res.Success, Error: res.Error}
```
Comment (abhimanyubabbar) [inline warehouse/admin.go:119]: Can we write it like this ?
`reply = &TestResponse{Status: res.Success, Error: res.Error}`

File: warehouse/admin.go [hunk 6]
```diff
@@ -77,30 +88,34 @@ func (wh *WarehouseAdmin) Query(s QueryInput, reply *warehouseutils.QueryResult)
...
-	res, err := tableUploadReq.GetWhTableUploads()
-	if err != nil {
-		return err
+	for _, v := range srcMap {
+		warehouse = v
+		break
 	}
-	bytes, err := json.Marshal(res)
+
+	pkgLogger.Infof(`[WH Admin]: Test warehouse: %s:%s`, warehouse.Type, warehouse.Destination.ID)
```
Comment (abhimanyubabbar) [inline warehouse/admin.go:108]: Can we update the log to be `pkgLogger.Infof([WH Admin]: Validating warehouse destination: %s:%s`

File: warehouse/admin.go [hunk 7]
```diff
@@ -77,30 +88,34 @@ func (wh *WarehouseAdmin) Query(s QueryInput, reply *warehouseutils.QueryResult)
...
+
+	destinationValidator := configuration_testing.NewDestinationValidator()
+	req := &configuration_testing.DestinationValidationRequest{Destination: warehouse.Destination}
+	res, err := destinationValidator.ValidateCredentials(req)
 	if err != nil {
-		return err
+		err = fmt.Errorf("unable to successfully validate destination: %s credentials, err: %v", warehouse.Destination.ID, err)
+		pkgLogger.Error(err)
```
Comment (abhimanyubabbar) [inline warehouse/admin.go:115]: Let's only log at the top level otherwise we would have duplicate logs. Wdyt ?

**Review by Jayachand — APPROVED** [review]
LGTM

**Pull request merged**